lxc-to-lxd/main_migrate.go

```diff
@@ -0,0 +1,602 @@
+			c.flagDryRun, c.flagRsyncArgs, c.flagDebug)
+		if err != nil {
+			fmt.Printf("Skipping container '%s': %v\n", container.Name(), err)
+			continue
+		}
+
+		// Delete container
+		if c.flagDelete {
+			if c.flagDryRun {
+				fmt.Println("Would destroy container now")
+			} else {
+				err := container.Destroy()
+				if err != nil {
+					fmt.Printf("Failed to destroy container '%s': %v\n", container.Name(), err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func validateConfig(conf []string, container *lxc.Container) error {
+	// Checking whether container has already been migrated
+	fmt.Println("Checking whether container has already been migrated")
+	if len(getConfig(conf, "lxd.migrated")) > 0 {
+		return fmt.Errorf("Container has already been migrated")
+	}
+
+	// Validate lxc.utsname / lxc.uts.name
+	value := getConfig(conf, "lxc.uts.name")
+	if value == nil {
+		value = getConfig(conf, "lxc.utsname")
+	}
+	if value == nil || value[0] != container.Name() {
+		return fmt.Errorf("Container name doesn't match lxc.uts.name / lxc.utsname")
+	}
+
+	// Validate lxc.aa_allow_incomplete: must be set to 0 or unset.
+	fmt.Println("Validating whether incomplete AppArmor support is enabled")
+	value = getConfig(conf, "lxc.apparmor.allow_incomplete")
+	if value == nil {
+		value = getConfig(conf, "lxc.aa_allow_incomplete")
+	}
+	if value != nil {
+		v, err := strconv.Atoi(value[0])
+		if err != nil {
+			return err
+		}
+
+		if v != 0 {
+			return fmt.Errorf("Container allows incomplete AppArmor support")
+		}
+	}
+
+	// Validate lxc.autodev: must be set to 1 or unset.
+	fmt.Println("Validating whether mounting a minimal /dev is enabled")
+	value = getConfig(conf, "lxc.autodev")
+	if value != nil {
+		v, err := strconv.Atoi(value[0])
+		if err != nil {
+			return err
+		}
+
+		if v != 1 {
+			return fmt.Errorf("Container doesn't mount a minimal /dev filesystem")
+		}
+	}
+
+	// Extract and valid rootfs key
+	fmt.Println("Validating container rootfs")
+	rootfs, err := getRootfs(conf)
+	if err != nil {
+		return err
+	}
+
+	if !shared.PathExists(rootfs) {
+		return fmt.Errorf("Couldn't find the container rootfs '%s'", rootfs)
+	}
+
+	return nil
+}
+
+func convertContainer(d lxd.ContainerServer, container *lxc.Container, storage string,
+	dryRun bool, rsyncArgs string, debug bool) error {
+	// Don't migrate running containers
+	if container.Running() {
+		return fmt.Errorf("Only stopped containers can be migrated")
+	}
+
+	fmt.Println("Parsing LXC configuration")
+	conf, err := parseConfig(container.ConfigFileName())
+	if err != nil {
+		return err
+	}
+
+	if debug {
+		fmt.Printf("Container configuration:\n  %v\n", strings.Join(conf, "\n  "))
+	}
+
+	// Check whether there are unsupported keys in the config
+	fmt.Println("Checking for unsupported LXC configuration keys")
+	keys := getUnsupportedKeys(getConfigKeys(conf))
+	for _, k := range keys {
+		if !strings.HasPrefix(k, "lxc.net.") &&
+			!strings.HasPrefix(k, "lxc.network.") &&
```

Comment:
Like, I thought we only allow e.g. `lxc.net.[i].*` and then specific subkeys like `hwaddr`. There's some filtering logic for this in `lxd/container_lxc.go`, I think.

Reply:
I thought the point of the check here is to skip lxc.net, lxc.network, lxc.cgroup and lxc.cgroup2 because we have special handling for those where we generate the relevant LXD config rather than dump that stuff into raw.lxc.